libbeat/processors/actions/append.go

```diff
@@ -0,0 +1,155 @@
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"fmt"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/processors"
+	"github.com/elastic/beats/v7/libbeat/processors/checks"
+	jsprocessor "github.com/elastic/beats/v7/libbeat/processors/script/javascript/module/processor"
+	conf "github.com/elastic/elastic-agent-libs/config"
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+type appendProcessor struct {
+	config appendProcessorConfig
+	logger *logp.Logger
+}
+
+type appendProcessorConfig struct {
+	Fields            []string      `config:"fields"`
+	TargetField       string        `config:"target_field"`
+	Values            []interface{} `config:"values"`
+	IgnoreMissing     bool          `config:"ignore_missing"`
+	IgnoreEmptyValues bool          `config:"ignore_empty_values"`
+	FailOnError       bool          `config:"fail_on_error"`
+	AllowDuplicate    bool          `config:"allow_duplicate"` //TODO: Add functionality to remove duplicate
+}
+
+func init() {
+	processors.RegisterPlugin("append_processor",
+		checks.ConfigChecked(NewAppendProcessor,
+			checks.RequireFields("target_field"),
+		),
+	)
+	jsprocessor.RegisterPlugin("AppendProcessor", NewAppendProcessor)
+}
+
+// NewAppendProcessor returns a new append_processor processor.
+func NewAppendProcessor(c *conf.C) (processors.Processor, error) {
+	config := appendProcessorConfig{
+		IgnoreMissing:     false,
+		IgnoreEmptyValues: false,
+		FailOnError:       true,
+		AllowDuplicate:    true,
+	}
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack the configuration of append processor: %w", err)
+	}
+
+	f := &appendProcessor{
+		config: config,
+		logger: logp.NewLogger("append_processor"),
+	}
+	return f, nil
+}
+
+func (f *appendProcessor) Run(event *beat.Event) (*beat.Event, error) {
+	var backup *beat.Event
+	if f.config.FailOnError {
+		backup = event.Clone()
+	}
+
+	err := f.appendValues(f.config.TargetField, f.config.Fields, f.config.Values, event)
+	if err != nil {
+		errMsg := fmt.Errorf("failed to append fields in append_processor processor: %w", err)
+		f.logger.Debug(errMsg.Error())
+		if f.config.FailOnError {
+			event = backup
+			if _, err := event.PutValue("error.message", errMsg.Error()); err != nil {
+				return nil, fmt.Errorf("failed to append fields in append_processor processor: %w", err)
+			}
+			return event, err
+		}
+	}
+
+	return event, nil
+}
+
+func (f *appendProcessor) appendValues(target string, fields []string, values []interface{}, event *beat.Event) error {
+	var arr []interface{}
+
+	val, err := event.GetValue(target)
+	if err != nil {
+		f.logger.Debugf("could not fetch value for key: %s. all the values will be appended in a new key %s.", target, target)
+	} else {
+		arr = append(arr, val)
```

Comment:
Do we need to check the type of `val` here?  I'm thinking we need to do the same as line 117 below.

Either way, it would be nice to add a test where target already exists as both a scalar and a list.

Reply:
I don't think a type check is required. as the `event.GetValue` returns `interface{}` and we will have to append if it's a scaler. We just have to iterate in case of a list. Hance the type check on L117.

Let me know if you think otherwise. 🤔

Reply:
You were right. I have added a type check in this line to avoid appending and array inside an array.